src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -0,0 +1,974 @@
+        num_chunks += extra_chunks;
+        return num_chunks;
+      } else {
+        current_group_span /= 2;
+      }
+    } else {
+      // This last group has fewer than _regular_group_size entries.
+      size_t chunk_span = current_group_span / _regular_group_size;
+      size_t last_group_size = unspanned_heap_size / chunk_span;
+      assert (last_group_size * chunk_span == unspanned_heap_size, "Chunks must precisely span regions");
+      num_chunks += last_group_size;
+      return num_chunks;
+    }
+  }
+  return num_chunks;
+}
+
+ShenandoahRegionChunkIterator::ShenandoahRegionChunkIterator(size_t worker_count) :
+    ShenandoahRegionChunkIterator(ShenandoahHeap::heap(), worker_count)
+{
+}
+
+ShenandoahRegionChunkIterator::ShenandoahRegionChunkIterator(ShenandoahHeap* heap, size_t worker_count) :
+    _heap(heap),
+    _regular_group_size(calc_regular_group_size()),
+    _first_group_chunk_size_b4_rebalance(calc_first_group_chunk_size_b4_rebalance()),
+    _num_groups(calc_num_groups()),
+    _total_chunks(calc_total_chunks()),
+    _index(0)
+{
+#ifdef ASSERT
+  size_t expected_chunk_size_words = _clusters_in_smallest_chunk * CardTable::card_size_in_words() * ShenandoahCardCluster::CardsPerCluster;
+  assert(smallest_chunk_size_words() == expected_chunk_size_words, "_smallest_chunk_size (" SIZE_FORMAT") is not valid because it does not equal (" SIZE_FORMAT ")",
+         smallest_chunk_size_words(), expected_chunk_size_words);
+#endif
+  assert(_num_groups <= _maximum_groups,
+         "The number of remembered set scanning groups must be less than or equal to maximum groups");
+  assert(smallest_chunk_size_words() << (_maximum_groups - 1) == _maximum_chunk_size_words,
+         "Maximum number of groups needs to span maximum chunk size to smallest chunk size");
+
+  size_t words_in_region = ShenandoahHeapRegion::region_size_words();
+  _region_index[0] = 0;
+  _group_offset[0] = 0;
+  if (words_in_region > _maximum_chunk_size_words) {
+    // In the case that we shrink the first group's chunk size, certain other groups will also be subsumed within the first group
+    size_t num_chunks = 0;
+    size_t effective_chunk_size = _first_group_chunk_size_b4_rebalance;
+    size_t  current_group_span = effective_chunk_size * _regular_group_size;
+    while (effective_chunk_size >= _maximum_chunk_size_words) {
+      num_chunks += current_group_span / _maximum_chunk_size_words;
+      effective_chunk_size /= 2;
+      current_group_span /= 2;
+    }
+    _group_entries[0] = num_chunks;
+    _group_chunk_size[0] = _maximum_chunk_size_words;
+  } else {
+    _group_entries[0] = _regular_group_size;
+    _group_chunk_size[0] = _first_group_chunk_size_b4_rebalance;
+  }
+
+  size_t previous_group_span = _group_entries[0] * _group_chunk_size[0];
+  for (size_t i = 1; i < _num_groups; i++) {
+    size_t previous_group_entries = (i == 1)? _group_entries[0]: (_group_entries[i-1] - _group_entries[i-2]);
```

Comment:
Sonar: Value stored to `previous_group_entries` during its initialization is never read

Reply:
https://bugs.openjdk.org/browse/JDK-8344321